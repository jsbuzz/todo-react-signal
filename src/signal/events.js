import { Event } from '../../react-signal/event-hive/event';

export const AddTodoEvent = class extends Event {
  constructor(title) {
    super();
    this.title = title;
  }
};
export const IdEvent = class extends Event {
  constructor(id) {
    super();
    this.id = id;
  }
};

export const ErrorEvent = class extends Event {
  constructor(error) {
    super();
    this.error = error;
  }
};

export const AddTodo = AddTodoEvent.withAlias('Todo:add');
export const RemoveTodo = IdEvent.withAlias('Todo:remove');
export const ToggleTodo = IdEvent.withAlias('Todo:toggle');
