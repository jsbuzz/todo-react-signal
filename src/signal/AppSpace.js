import {
  NameSpace,
  InitState,
  set,
} from '../../react-signal/event-hive/namespace';
import { UpdateActive } from './events';

const AppSpace = NameSpace.schema('Ns.App', () => ({
  activeTodos: [
    InitState, set(0),
    
    // update number of active todos
    UpdateActive, activeTodos => ({ active }) => console.log(active) || active
  ]
}));

window.AppSpace = AppSpace;

export default AppSpace;
