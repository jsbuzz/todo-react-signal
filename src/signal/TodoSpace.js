import {
  NameSpace,
  InitState,
  set,
  modify
} from '../../react-signal/event-hive/namespace';
import { AddTodo, RemoveTodo, UpdateTodo, RestoreTodos } from './events';

const TodoSpace = NameSpace.schema((todoId = 0) => ({
  todos: [
    InitState, set(new Map()),
    
    // Add new todo
    AddTodo, modify(todos => ({ title }) => {
      const id = ++todoId;
      const todo = { id, title, done: false };

      todos.set(id, todo);
    }),
    
    // restore todos from cache
    RestoreTodos, modify(todos => ({ savedTodos }) => {
      savedTodos.forEach(todo => {
        todoId = todo.id;
        todos.set(todo.id, todo);
      });
    }),
    
    // remove a todo by id
    RemoveTodo, modify(todos => ({ id }) => todos.delete(id)),
    
    // update a todo by id
    UpdateTodo, modify(todos => ({ todo }) => todos.set(todo.id, todo))
  ]
}));

window.TodoSpace = TodoSpace;

export default TodoSpace;
