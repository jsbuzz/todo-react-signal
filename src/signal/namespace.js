import { NameSpace, InitState } from '../../react-signal/event-hive/namespace';
import { AddTodo, RemoveTodo, ToggleTodo } from './events';

const TodoSpace = NameSpace.get('Ns.Todos');

const findTodo = (todos, todoId) => {
  let i = 0;
  for (const todo of todos) {
    if (todo.id === todoId) {
      return i;
    }
    i += 1;
  }
  return -1;
};

const toggleTodo = (todo) => {
  todo.done = !todo.done;
}

let todoId = 0;
TodoSpace.defineState({
  todos: state => [
    InitState, () => (state.todos = []),
    AddTodo, ({ title }) => (state.todos.push({
      id: ++todoId,
      title,
      done: false,
    })),
    RemoveTodo, ({ id }) => (state.todos.splice(findTodo(state.todos, id), 1)),
    ToggleTodo, ({ id }) => (toggleTodo(state.todos[findTodo(state.todos, id)])),
  ],
});

export default TodoSpace;
