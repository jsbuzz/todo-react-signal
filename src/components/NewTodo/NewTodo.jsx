import React from 'react';
import { AddTodo } from '../../signal/events';

const NewTodo = (_, namespace) => (
  <input
    autoFocus="autofocus"
    autoComplete="off"
    placeholder="What needs to be done?"
    className="new-todo"
    onKeyUp={keyup => {
      const { currentTarget, keyCode } = keyup;
      if (keyCode === 13 && currentTarget.value.length) {
        namespace().trigger(AddTodo.with(currentTarget.value));
        currentTarget.value = '';
      }
    }}
  />
);

export default NewTodo;
