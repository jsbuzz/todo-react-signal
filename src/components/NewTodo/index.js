import Connect from '../../../react-signal';
import NewTodo from './NewTodo';

export default Connect(NewTodo);
