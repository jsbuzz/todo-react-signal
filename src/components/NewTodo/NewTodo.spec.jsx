import React from 'react';
import { shallow } from 'enzyme';
import Connect, { TestNameSpace } from '../../../react-signal/test-connect';

import { AddTodo } from '../../signal/events';
import NewTodo from './NewTodo';

describe('components/NewTodo', () => {
  let NS;
  let component;

  beforeEach(() => {
    NS = new TestNameSpace();
    const ConnectedComponent = Connect(NewTodo, NS);
    component = shallow(<ConnectedComponent />);
  });

  it('should match snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  describe('WHEN enter is pressed', () => {
    describe('AND there is NO title set', () => {
      it('should NOT trigger any event', async () => {
        const input = component.find('input');
        input.simulate('keyUp', { keyCode: 13, currentTarget: { value: '' }});

        expect(NS.lastEvent).toBe(null);
      });
    });

    describe('AND there is a title set', () => {
      const todoTitle = 'Test Title';

      it('should trigger AddTodo with the correct title', () => {
        const input = component.find('input');
        const currentTarget = { value: todoTitle };
        input.simulate('keyUp', { keyCode: 13, currentTarget });

        expect(NS.lastEvent).toBeInstanceOf(AddTodo);
        expect(NS.lastEvent.title).toEqual(todoTitle);
      });
    });
  });
});