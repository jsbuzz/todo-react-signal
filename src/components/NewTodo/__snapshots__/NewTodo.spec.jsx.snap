// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/NewTodo should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NewTodo />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "autoComplete": "off",
      "autoFocus": "autofocus",
      "className": "new-todo",
      "onKeyUp": [Function],
      "placeholder": "What needs to be done?",
    },
    "ref": null,
    "rendered": null,
    "type": "input",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "autoComplete": "off",
        "autoFocus": "autofocus",
        "className": "new-todo",
        "onKeyUp": [Function],
        "placeholder": "What needs to be done?",
      },
      "ref": null,
      "rendered": null,
      "type": "input",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
